include/revng/Pipes/InvalidationHelpers.h

```diff
@@ -0,0 +1,152 @@
+  const auto Location = location(revng::ranks::DynamicFunction, Entry);
+  const auto &Related = CR->Relations.at(Location.toString()).Related;
+  for (const yield::RelationTarget &RT : Related) {
+    if (RT.Kind == yield::RelationType::IsDynamicallyCalledFrom) {
+      auto Caller = genericLocationFromString<0>(RT.Location,
+                                                 revng::ranks::Function,
+                                                 revng::ranks::BasicBlock,
+                                                 revng::ranks::Instruction);
+      revng_assert(Caller.has_value());
+      Result.insert(std::get<0>(*Caller));
+    }
+  }
+}
+
+// Invalidate all the targets belonging to function rank
+inline void
+invalidateAllTargetsPerFunctionRank(const TupleTree<model::Binary> &Model,
+                                    std::set<MetaAddress> &Result) {
+  for (const model::Function &Function : Model->Functions)
+    Result.insert(Function.Entry);
+}
+
+namespace detail {
+
+inline void
+insertCallersAndTransitiveClosureInternal(const TupleTree<CrossRelations> &CR,
+                                          const TupleTree<model::Binary> &Model,
+                                          const std::string &Location,
+                                          std::set<MetaAddress> &Result) {
+  auto CRGraph = CR->toCallGraph();
+  using CRNode = typename decltype(CRGraph)::Node;
+
+  const CRNode *EntryNode = nullptr;
+  for (const CRNode *Node : CRGraph.nodes())
+    if (Node->data() == Location)
+      EntryNode = Node;
+
+  revng_assert(EntryNode != nullptr);
+
+  // Inverse DFS visit that aims at finding its callers, and its
+  // transitive closure, if they are all `Inline`.
+  llvm::df_iterator_default_set<const CRNode *> NoInlineFunctions;
+  auto VisitRange = llvm::inverse_post_order_ext(EntryNode, NoInlineFunctions);
+
+  for (const CRNode *Node : VisitRange) {
+    namespace p = pipeline;
+    namespace ranks = revng::ranks;
+
+    for (const CRNode *From : Node->predecessors()) {
+      auto MaybeKey = p::genericLocationFromString<0>(From->data(),
+                                                      ranks::Function,
+                                                      ranks::BasicBlock,
+                                                      ranks::Instruction);
+
+      if (MaybeKey.has_value()) {
+        auto &Function = Model->Functions.at(std::get<MetaAddress>(*MaybeKey));
+        if (Function.Attributes.count(model::FunctionAttribute::Inline) == 0)
+          NoInlineFunctions.insert(From);
+      }
+    }
+
+    auto Address = p::genericLocationFromString<0>(Node->data(),
+                                                   ranks::Function,
+                                                   ranks::BasicBlock,
+                                                   ranks::Instruction);
+    if (Address.has_value())
+      Result.insert(std::get<MetaAddress>(*Address));
+  }
+}
```

Comment:
Can't you simply pre-populate `NoInlineFunctions` with all the functions that are not inline (except for the starting point)?

Reply:
What's the point? It should be way more optimal to rely on `inverse_post_order_ext` to both traversal and filtering then first manually gathering non-inline functions (iterating over all the functions) and then do a conditional traversal. At least, that ought to be profiled to confirm which one of the two approaches is faster.

Reply:
If `Inline` attribute is rare, it might be better to just gather a set of all the inline functions instead of listing the non-inline ones, and then the inner loop would just be a lookup on that set?

Reply:
We're not discussing performance, we're discussing simplicity.
Using `_ext` without manually changing the set of visited nodes in the loop is much more intuitive and simpler to explain: "we mark all non-inline nodes as visited and then perform a simple inverse dfs".

If you're concerned about performance, you can reuse the initial set among various invocations of this method (copying it).

Reply:
Following your direction, pre-populating `NoInlineFunctions` implies that, no sooner `inverse_post_order_ext ` later sees a node that is already in the set, it abandons that branch of the traversal, not exploring further. If the callers of the entry function are added to the set `NoInlineFunctions`, the inverse POT won't visit them. This means we end up getting neither the transitive closure nor the callers. To make it work, we would need to use the complementary set, namely, something like:
```diff
+  llvm::df_iterator_default_set<const CRNode *> AnotherSet;
-  auto VisitRange = llvm::inverse_post_order_ext(EntryNode, NoInlineFunctions);
+  auto VisitRange = llvm::inverse_post_order_ext(EntryNode, AnotherSet);

  for (const CRNode *Node : VisitRange) {
+    if (!llvm::is_contained(Node, NoInlineFunctions)) {
+      AnotherSet.emplace(Node);
+      continue;
+    }
    using namespace pipeline;
    namespace ranks = revng::ranks;

    auto Address = genericLocationFromString<0>(Node->data(),
                                                ranks::Function,
                                                ranks::BasicBlock,
                                                ranks::Instruction);
    if (Address.has_value())
      Result.insert(std::get<MetaAddress>(*Address));
  }
}
```
This seemingly adds further complexity. Personally, I would refrain from doing such a change.

Reply:
It's more straightforward to read as opposed to "peeking" into successor's list while visiting nodes.
Also, you can pre-create once and for all the set of non-`Inline` functions and just copy it every time, so step 2) is not even there.

I already expressed these observations [in a previous comment](https://github.com/revng/revng/pull/209#discussion_r1029355688).